### ISSUE TITLE
Resize of root disk doesn't work on Alpine 3.12.1 aarch64

### ISSUE THREAD

**Issue by tsarna** [mcrute/tiny-ec2-bootstrap#15]
I created a VM using the new aarch64 AMIs for Alpine 3.12.1 with a 4GB root disk, and just noticed that the root filesystem is only ~1G.

I think the issue is the different partitioning on ARM. On amd64, the root disk is just /dev/nvme0n1, but on ARM I see:

```
/dev/nvme0n1p2          985.0M    167.0M    751.1M  18% /
/dev/nvme0n1p1            5.0M    278.0K      4.7M   5% /boot/efi
```

I think resize2fs isn't able to expand the filesystem since it's already occupying the whole partition, and when I duplicate the resize command that the bootstrap runs:

```
$ resize2fs /dev/nvme0n1p2
resize2fs 1.45.6 (20-Mar-2020)
The filesystem is already 260347 (4k) blocks long.  Nothing to do!
```

So I think the resize step needs to check if the root fs is a partition or raw disk, and if a partition needs to repartition.

**Comment by tomalok** [comment]
Thanks for bringing this to our attention, I'll take a look into this soon.

**Comment by tsarna** [comment]
The easy way to fix this manually was to just `growpart /dev/nvme0n1 2` first and then resize. It seems a bit trickier to package this as something working in the AMIs though, especially since the cloud-utils-growpart package is only available on edge.

**Comment by tomalok** [comment]
The root of the issue is that the **x86_64** AMI does not partition the EBS volume...
```
alpine@ip-172-31-0-161:~$ df -h
Filesystem                Size      Used Available Use% Mounted on
devtmpfs                 10.0M         0     10.0M   0% /dev
shm                     233.5M         0    233.5M   0% /dev/shm
/dev/nvme0n1              9.8G    111.6M      9.3G   1% /
tmpfs                    46.7M     96.0K     46.6M   0% /run
df: /sys/kernel/debug/tracing: Permission denied
alpine@ip-172-31-0-161:~$ mountpoint -n /
/dev/nvme0n1 /
alpine@ip-172-31-0-161:~$ uname -a
Linux ip-172-31-0-161 5.4.72-0-virt #1-Alpine SMP Mon, 19 Oct 2020 06:22:29 UTC x86_64 Linux
```
...but the **aarch64** AMI ***partitions*** it...
```
alpine@ip-172-31-0-62:~$ df -h
Filesystem                Size      Used Available Use% Mounted on
devtmpfs                 10.0M         0     10.0M   0% /dev
shm                     225.9M         0    225.9M   0% /dev/shm
/dev/nvme0n1p2          985.0M    115.9M    802.2M  13% /
tmpfs                    45.2M     96.0K     45.1M   0% /run
df: /sys/kernel/debug/tracing: Permission denied
/dev/nvme0n1p1            5.0M    278.0K      4.7M   5% /boot/efi
alpine@ip-172-31-0-62:~$ mountpoint -n /
/dev/nvme0n1p2 /
alpine@ip-172-31-0-62:~$ sudo fdisk /dev/nvme0n1
Found valid GPT with protective MBR; using GPT

Command (m for help): p
Disk /dev/nvme0n1: 20971520 sectors, 2048M
Logical sector size: 512
Disk identifier (GUID): f02527ff-ee99-4da4-bbf9-210c0fee36fa
Partition table holds up to 128 entries
First usable sector is 2048, last usable sector is 2097118

Number  Start (sector)    End (sector)  Size Name
     1            4096           14335 5120K
     2           14336         2097118 1016M

Command (m for help): q

alpine@ip-172-31-0-62:~$ uname -a
Linux ip-172-31-0-62 5.4.72-0-virt #1-Alpine SMP Mon, 19 Oct 2020 06:22:29 UTC aarch64 Linux
```

I'm not sure (@mcrute may know better) but separate partitions *may* be a requirement of `grub` which is used as the bootloader for **aarch64**.

`resize2fs` on **x86_64** queries the EBS volume for its size, whereas it queries the partition table (inherited from the image) on **aarch64**.

Somehow fixing this over in github.com/alpine-ec2-ami would mean we could just roll new AMI revisions for affected non-EOL Alpine releases.

Easiest fix would be to install grub on a raw EBS volume -- if that's even possible?

Back a while ago, I was taking a look at the `u-boot` bootloader for **aarch64**, I seem to recall it had some compatibility with the `syslinux` bootloader that we use for **x86_64** -- if we could get that working in lieu of `grub`, we'd reduce the footprint of the **aarch64** AMI, and move back to a single volume solution...

**Comment by tomalok** [comment]
Another way?  (It appears that partitioning may be a BIOS/EFI requirement for **aarch64**...)

Our AMI's root partition is guaranteed to be the last one in the image -- we may be able to tweak it with `sfdisk` before doing the `resize2fs`...  I'm going to try http://karelzak.blogspot.com/2015/05/resize-by-sfdisk.html

(FWIW, adding `cloud-util-growpart` to the current AMI installs a total of 27 packages, and an extra 8 MiB or so...  `sfdisk` adds 3 packages and ~1 MiB.)

**Comment by tsarna** [comment]
growpart was the first thing I turned up when trying to figure out how to do it manually without having to calculate partition sizes and offsets by hand. If there's a better/smaller tool I'm all for it!

**Comment by tomalok** [comment]
```
root@ip-172-31-49-138:~# echo ", +" | sfdisk --no-reread -N 2 /dev/nvme0n1
GPT PMBR size mismatch (2097151 != 20971519) will be corrected by write.
The backup GPT table is not on the end of the device. This problem will be corrected by write.
Disk /dev/nvme0n1: 10 GiB, 10737418240 bytes, 20971520 sectors
Disk model: Amazon Elastic Block Store
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: 00E78CAF-AFC7-49A8-A758-31719F801C98

Old situation:

Device         Start     End Sectors  Size Type
/dev/nvme0n1p1  4096   14335   10240    5M EFI System
/dev/nvme0n1p2 14336 2097118 2082783 1017M Linux filesystem

/dev/nvme0n1p2:
New situation:
Disklabel type: gpt
Disk identifier: 00E78CAF-AFC7-49A8-A758-31719F801C98

Device         Start      End  Sectors Size Type
/dev/nvme0n1p1  4096    14335    10240   5M EFI System
/dev/nvme0n1p2 14336 20971486 20957151  10G Linux filesystem

The partition table has been altered.
Calling ioctl() to re-read partition table.
Re-reading the partition table failed.: Resource busy
The kernel still uses the old table. The new table will be used at the next reboot or after you run partprobe(8) or partx(8).
Syncing disks.
```
```
root@ip-172-31-49-138:~# resize2fs /dev/nvme0n1p2
resize2fs 1.45.6 (20-Mar-2020)
The filesystem is already 260347 (4k) blocks long.  Nothing to do!
```
```
root@ip-172-31-49-138:~# partprobe /dev/nvme0n1
partprobe: /dev/nvme0n1: Resource busy
```
Needed to **reboot**...  Unless someone knows any tricks?
```
root@ip-172-31-49-138:~# resize2fs /dev/nvme0n1p2
resize2fs 1.45.6 (20-Mar-2020)
Filesystem at /dev/nvme0n1p2 is mounted on /; on-line resizing required
old_desc_blocks = 1, new_desc_blocks = 1
The filesystem on /dev/nvme0n1p2 is now 2619643 (4k) blocks long.
```
```
root@ip-172-31-49-138:~# df -h
Filesystem                Size      Used Available Use% Mounted on
devtmpfs                 10.0M         0     10.0M   0% /dev
shm                     225.8M         0    225.8M   0% /dev/shm
/dev/nvme0n1p2            9.8G    113.0M      9.3G   1% /
tmpfs                    45.2M     92.0K     45.1M   0% /run
/dev/nvme0n1p1            5.0M    278.0K      4.7M   5% /boot/efi
```

**Comment by tomalok** [comment]
Note the...
```
GPT PMBR size mismatch (2097151 != 20971519) will be corrected by write.
The backup GPT table is not on the end of the device. This problem will be corrected by write.
```
Looks like `sfdisk` is also doing the right thing with regards to the backup GPT.

**Comment by tsarna** [comment]
Hmm, I have not needed to reboot between using growpart and resize2fs, I wonder what it does differently?

**Comment by tomalok** [comment]
`growpart` may use `partx` which is part of `util-linux` -- which is a dependency, IIRC.

**Comment by mcrute** [comment]
> I'm not sure (@mcrute may know better) but separate partitions _may_ be a requirement of `grub` which is used as the bootloader for **aarch64**.

A GPT partition table and a FAT32 volume for the boot loader is a requirement for UEFI (per the spec). The firmware knows how to find the boot partition based on a GPT flag and then how to read it to find its bootloader based on filesystem naming conventions.

**Comment by mcrute** [comment]
> `growpart` may use `partx` which is part of `util-linux` -- which is a dependency, IIRC.

`growpart` is [definitely using partx](https://git.launchpad.net/cloud-utils/tree/bin/growpart#n442) behind the scenes after it finishes growing the partition. Unfortunately, the way that Alpine packages `partx` it's not separable from `util-linux`, which is bringing along the bulk of the dependencies @tomalok noted.

I'm no longer opposed to pulling `growpart` in as a dependency. The dependency graph has been pruned a lot from what it was when I originally rejected the idea when writing the bootstrap. The thing that really holds me back right now is that it's only in edge and appears to be part of a larger restructuring effort on those packages.

In the near term I think we should add a dependency on `util-linux` to the bootstrap and duplicate the critical `growpart` logic here but keep it contained enough that when `growpart` lands in a numbered release we can remove our duplicated code in preference to using it.

**Comment by tomalok** [comment]
What adverse effects would there be if we did the `sfdisk` resizing (as early as possible), and then if we needed to, rebooted?
  This would be before touching the "we already bootstrapped" breadcrumb file, of course.

**Comment by mcrute** [comment]
I'm not keen on rebooting during the boot process because it will take longer for the machine to become ready for use. Is there a reason you'd rather not pull in `util-linux` and do what `growpart` is doing? If it's a dependency graph concern we could try to split `partx` into a separate package from `util-linux`. The entire graph for just the `partx` binary looks like just 3 packages and `musl` is one of them.

**Comment by tomalok** [comment]
Splitting `partx` to get just what we need would be a good idea -- even if it doesn't get backported, we can do what we did for the IPv6 stuff (**edge** and future versions use `iproute2-minimal`, but **3.12** and earlier need to use the full `iproute2`).

I'm guessing a split would basically be an independently-installable subpackage `util-linux-partx`, that also gets pulled in when `util-linux` is added to a system?

**Comment by mcrute** [comment]
> Splitting `partx` to get just what we need would be a good idea -- even if it doesn't get backported, we can do what we did for the IPv6 stuff (**edge** and future versions use `iproute2-minimal`, but **3.12** and earlier need to use the full `iproute2`).
> 
> I'm guessing a split would basically be an independently-installable subpackage `util-linux-partx`, that also gets pulled in when `util-linux` is added to a system?

I'd probably just name it `partx` and make it a dependency of `util-linux`, similar to `blkid` and `cfdisk`. 

> Can the dependencies of `tiny-ec2-bootstrap` be declared such that "use `util-linux`, unless `util-linux-partx` is available"?

The only way I'm aware that this could work would be some metapackage tricks and package backports that I feel pretty confident upstream would thumbs-down.

Another option for older releases would be to rip out the parts of `partx` that we need and vendor them as a binary with this package. Looking at the source those bits are pretty self-contained and would only pull in a dependency on `libblkid`. edge would use the new package upstream and older releases would use the vendored version. What do you think?

**Comment by dermotbradley** [comment]
> I'm no longer opposed to pulling growpart in as a dependency. The dependency graph has been pruned a lot from what it was when I originally rejected the idea when writing the bootstrap. The thing that really holds me back right now is that it's only in edge and appears to be part of a larger restructuring effort on those packages.

@mcrute Alpine cloud-utils package maintainer here. The restructuring of cloud-utils is basically finished - I split it up specifically to get reduced dependancies when installing growpart as I use it with cloud-init. Yes its only in Edge currently but 3.13 will be out fairly soon (I noticed ncopa on IRC yesterday talking about getting the 3.13 builders ready).

I've just raised a MR, https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/14512, to reduce the dependancies/size of cloud-utils-growpart further - in the recent upstream 0.32 release growpart added a call to "udevadm settle", this MR has a patch to make this optional and I've therefore removed the package dependancy on "eudev". Hopefully the MR will get merged in the next day or so (i.e. in time for 3.13).

@tomalok the size when installing cloud-utils-growpart should drop by maybe 2Mb (eudev alone is 1.16M on aarch64) and likewise quite a few dependancies pulled in by eudev will disappear.

**Comment by tomalok** [comment]
I've submitted a merge request to split out `partx` from `util-linux` -- https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/14558

**Comment by dermotbradley** [comment]
> I've submitted a merge request to split out `partx` from `util-linux` -- https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/14558

Good idea! I see its literally just been merged so now I need to update cloud-utils-growport to switch the util-linux dependency for partx.

Seperately I decided to do an audit of the growpart code and realised that there's no actual need for the "coreutils" dependancy - the Busybox versions of the various utilities all appear to be compatible so I'm removing the "coreutils" dependency.

I'm currently doing the same audit for its dependency of "util-linux" - if this can be also removed then, in addition to the above changes, the package would be left with a list of just "blkid", "partx", "sfdisk".

I'll update this note once I've finished testing and submitted an MR...

**Comment by tomalok** [comment]
We'll need to get the next release of `tiny-ec2-bootstrap` sorted soon to get it into *3.13*.  I'll work on a PR this afternoon/evening.

**Comment by tomalok** [comment]
Found and fixed issue #17 when backporting #16 to the 3.12.1 AMIs releases.

The aarch64 AMI for 3.12.1 on https://github.com/mcrute/alpine-ec2-ami/blob/master/releases/README.md should properly resize the root partition now.  (The `util-linux` `mountpoint` didn't support `-n` like `busybox`'s does...)
